pol-core/ecompile/ECompileMain.cpp

```diff
@@ -713,6 +820,250 @@ void parallel_process( const std::set<std::string>& files )
+    tmp += fmt::format( "  prune cache (sel): {}\n",
+                        (long long)summary.profile.prune_cache_select_micros / 1000 );
+    tmp += fmt::format( "  prune cache (del): {}\n",
+                        (long long)summary.profile.prune_cache_delete_micros / 1000 );
+    tmp += "\n";
+    tmp += fmt::format( "      - ambiguities: {}\n", (long)summary.profile.ambiguities );
+    tmp += fmt::format( "       - cache hits: {}\n", (long)summary.profile.cache_hits );
+    tmp += fmt::format( "     - cache misses: {}\n", (long)summary.profile.cache_misses );
+  }
+
+  INFO_PRINTLN( tmp );
+}
+
+void EnterWatchMode()
+{
+  std::list<WatchFileMessage> watch_messages;
+  std::set<fs::path> to_compile;
+  efsw::FileWatcher fileWatcher;
+  EfswFileWatchListener listener( fileWatcher,
+                                  std::set<fs::path>{ ".src", ".hsr", ".asp", ".inc", ".em" } );
+
+  auto handle_compile_file = [&]( const fs::path& filepath )
+  {
+    // Existing file, compile all owners dependent on it.
+    if ( auto itr = dependency_owners.find( filepath ); itr != dependency_owners.end() )
+    {
+      to_compile.insert( itr->second.begin(), itr->second.end() );
+    }
+    // New file, compile only this file (if it is compilable).
+    else
+    {
+      auto ext = filepath.extension();
+      if ( ext.compare( ".src" ) == 0 || ext.compare( ".hsr" ) == 0 || ext.compare( ".asp" ) == 0 )
+      {
+        to_compile.emplace( filepath );
+      }
+    }
+  };
+
+  auto handle_delete_file = [&]( const fs::path& filepath )
+  {
+    to_compile.erase( filepath );
+    if ( auto itr = owner_dependencies.find( filepath ); itr != owner_dependencies.end() )
+    {
+      // For all dependencies, remove the deleted file from set of owners for
+      // that dependency.
+      for ( const auto& depnamepath : itr->second )
+      {
+        // If dependency exists in map of dependency -> owners
+        if ( auto owners_itr = dependency_owners.find( depnamepath );
+             owners_itr != dependency_owners.end() )
+        {
+          // Erase filepath from set of owners
+          owners_itr->second.erase( filepath );
+
+          // If owners set is empty, this dependency is no longer used by
+          // anything.
+          if ( owners_itr->second.empty() )
+          {
+            //  Erase from map of dependency -> owners.
+            dependency_owners.erase( owners_itr );
+            listener.remove_file( depnamepath );
+            VERBOSE_PRINTLN( "Remove watch file {}", depnamepath );
+          }
+        }
+      }
+      owner_dependencies.erase( itr );
+    }
+  };
+
+  auto add_dir = [&]( const std::string& elem )
+  {
+    fs::path dir( elem );
+    if ( fs::exists( dir ) )
+    {
+      auto dirpath = fs::canonical( dir );
+      if ( listener.add_watch_dir( dirpath ) )
+      {
+        VERBOSE_PRINTLN( "Add watch dir {}", dirpath );
+      }
+    }
+  };
+
+  for ( const auto& dep_owners : dependency_owners )
```

Comment:
I think always empty at start

Reply:
`dependency_owners` gets filled during compilation, and then get watched in `EnterWatchMode`.